### PR TITLE
Tactical cleanup 2.5.0

### PR DESCRIPTION
**Pull request by nk3nny**
**This pull request contains bugfixes and changes to the tactical ability of the AI.** 
- [x] Next move is to  add a new function (assaultSympathetic.sqf) 
- [x] Rename "lambs_danger_cqb_pos" variable to something more sensible. 
- [x] Fix check bodies state in civilian danger.fsm
- [x] Re-implement reinforcement system
- [x] Test run on Tuesday

**Current changes**: 

brain.sqf
 - Improved simplified gesture and share information clause

brainAssess.sqf
 - Fixed suppression no longer uses eyePos
 - Added tactical improvements to sympathetic assaults
 - Added debug information to sympathetic assaults (will be put in unique function)
 - Added speed adjustment to default formation speeds
 - Added cover movement again

brainEngage.sqf
 - Improved timings
 - Fixed suppression no longer uses eyePos

brainForced
 - Added speed tweak for "ATTACK" commanded units

doAssault.sqf
 - Added forced improved pathfinding

doFleeing.sqf
 - Added debug information

doGroupAssault.sqf
 - Improved timings
 - Improved movement speed settings

doGroupFlank.sqf
 - Improved stance options

isForced.sqf
 - Added check for "CARELESS" behaviour

isLeader
 - Tightnened requirements for leader checks 
 - Added small chance non-leaders will make tactical calls

shareinformation.sqf
 - Fixed 'isNull' error in function
 - Improved debug markers

tactics.sqf
 - Added 'known enemy' paramater

tacticsAssault.sqf
 - Improved timings
 - Added check for distance when using UGL-smoke and smoke grenades
 - Added debug markers

tacticsAssess.sqf
 - Improved readability
 - Added missing variable checks 
 - Fixed 'enemy far away or above' tactics
 - Fixed callouts on small units 
 - Many improvements to tactics options

tacticsContact.sqf
 - Added known enemy parameter
 - Changed default communication range is now SHOUT --> full communications happen later
 - Various clean ups to code
 
tacticsFlank.sqf
 - Added debug markers
 - minor tweaks 

tacticsGarrison.sqf
 - Improved timings
 - Added units rallying after garrison
 - Added debug markers

tacticsHide.sqf
 - Added debug markers

tacticsSuppress.sqf
 - Added debug markers

vehicle.sqf 
 - Fixed suppression no longer uses eyePos

lambs_danger.fsm
 - Changed Dodges no longer add 'Can Fire' to _queue
 - Removed Assess state interrupting on filled _queue
 - Fixed various function calls

lambs_danger_dangerUnit
 - changed default value from 25 to 15 (testing!)

### CODE REVIEW

**Review by jokoho48 — CHANGES_REQUESTED** [review]

File: addons/danger/functions/fnc_doAssaultMemory.sqf [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * Example:
+ * [bob] call lambs_danger_fnc_assaultMemory;
+ *
+ * Public: No
+*/
+params ["_unit", ["_groupMemory", []]];
+
+// check it
+if (_groupMemory isEqualTo []) then {
+    _groupMemory = (group _unit) getVariable [QGVAR(groupMemory)];
```
Comment (jokoho48) [inline addons/danger/functions/fnc_doAssaultMemory.sqf:22]: store group in varaible

File: addons/danger/functions/fnc_shareInformation.sqf [hunk 2]
```diff
@@ -48,17 +47,22 @@ private _groups = allGroups select {
...
-// share information
 {
+    // share information
     if !(isNull _target) then {
         [_x, [_target, _knowsAbout min GVAR(maxRevealValue)]] remoteExec ["reveal", leader _x];
     };
 
+    // reinforce
+    if ((_x getVariable [QGVAR(enableGroupReinforce), false]) && { (_x getVariable [QGVAR(enableGroupReinforceTime), -1]) < time}) then {
+        [_x, [getpos _unit, (_unit targetKnowledge _target) select 6] select (_knowsAbout > 0.5)] remoteExec ["move", leader _x];
```
Comment (jokoho48) [inline addons/danger/functions/fnc_shareInformation.sqf:58]:
```suggestion
        [_x, [getPos _unit, (_unit targetKnowledge _target) select 6] select (_knowsAbout > 0.5)] remoteExec ["move", leader _x];
```

File: addons/danger/functions/fnc_tacticsAssess.sqf [hunk 3]
```diff
@@ -70,72 +75,91 @@ if !(_enemies isEqualTo []) then {
...
+        // active or passive
+        if (RND(0.5)) then {
+            _plan append [TACTICS_SUPPRESS, TACTICS_SUPPRESS, TACTICS_HIDE];
+            if ((combatMode _unit) isEqualTo "RED" || {speedMode _unit isEqualTo "FULL"}) then {_plan pushBack TACTICS_ASSAULT;};
+            _pos = _unit getHideFrom (_enemies select _target);
+        } else {
+            _plan append [TACTICS_GARRISON, TACTICS_GARRISON];
+            _pos = getPos _unit;
+        };
     };
```
[the remark this answers is not in the document]
Comment (nk3nny) [reply addons/danger/functions/fnc_tacticsAssess.sqf:115]: Have gone through and given individual names for each target list. Perhaps increases legibility?

File: addons/danger/functions/fnc_tacticsAssess.sqf [hunk 4]
```diff
@@ -70,72 +75,91 @@ if !(_enemies isEqualTo []) then {
...
-    if (_targets != -1 && {!GVAR(disableAIAutonomousManoeuvres)}) exitWith {
-        _plan append [2, 2, 3];    // garrison, garrison, assault
-        _pos = [_unit getHideFrom (_enemies select _targets), getPosASL _unit] select _inside;
+    if (_target != -1) exitWith {
+        _plan append [TACTICS_GARRISON, TACTICS_ASSAULT, TACTICS_ASSAULT];
+        _pos = [_unit getHideFrom (_enemies select _target), getPosASL _unit] select _inside;
     };
 
     // inside? stay safe
     if (_inside) exitWith {_plan = [];};
```
[the remark this answers is not in the document]
Comment (nk3nny) [reply addons/danger/functions/fnc_tacticsAssess.sqf:97]: Ehh, No? :) 

So the tacticsAsssess state will go through a long set of evaluations of whether to move out of the building or not. There is a larger selection of potential triggers than before. Might be it? *shrug*

File: addons/danger/functions/fnc_tacticsContact.sqf [hunk 5]
```diff
@@ -44,10 +46,14 @@ _group setVariable [QGVAR(tacticsTask), "In contact!", EGVAR(main,debug_function
             _group setVariable [QGVAR(isExecutingTactic), nil];
             _group setVariable [QGVAR(tacticsTask), nil];
             _group enableAttack _enableAttack;
+            private _leader = leader _group;
+            if (_leader call EFUNC(main,isAlive) && {_leader call FUNC(isLeader)}) then {
```
Comment (jokoho48) [inline addons/danger/functions/fnc_tacticsContact.sqf:50]: as far as i can see isLeader already Checks if _leader is alive, so the first call would be redundant.

Comment (nk3nny) [reply addons/danger/functions/fnc_tacticsContact.sqf]: Good catch! Fixed.

**Pull request merged**